docs/proposals/connector-protocol-v3.md

```diff
@@ -462,42 +465,61 @@ NOTE: For both multi-worker Wallaroo clusters and the in-progress `parallelism`
+
+A MESSAGE frame that contains a 2PC Phase 1 request will specify the byte offset range(s) of the data sent to the sink for the logical time range that is managed by the Wallaroo global checkpoint protocol.
+
+The connector sink should be able to manage the durability of the data in this byte offset range(s).  If the connector sink cannot manage durability of this sink data, then the connector sink must always vote COMMIT during Phase 1; such a sink then cannot provide the end-to-end message delivery guarantees that the connector sink protocol is intended to provide.  In such cases, we recommend that the Wallaroo TCPSink be used instead.
+
+#### 5.a When the connector sink received a 2PC Phase 1 request
+
+If the sink choose to reply with a Phase 1 COMMIT, then the sink is responsible for buffering all data received since the last 2PC round.  The sink data cannot be re-sent by Wallaroo if the sink is careless and destroys/loses/corrupts the data that it has received.
+
+If the sink chooses to reply with a Phase 1 ABORT message, then the sink has no responsibility for the data received since the last 2PC round.  Wallaroo must abort the entire checkpoint in response to this ABORT message, and, therefore, Wallaroo must send a Phase 2 ABORT decision to all sinks.
+
+#### 5.b. When the connector sink receives a 2PC Phase 2 request
+
+If the Phase 2 decision is COMMIT, then the connector sink must make the application sink data in the transaction's byte range(s) durable, using whatever app-specific means necessary.  For example:
+* For file I/O, to write to a file(s) + flush(es) + fsync(s), etc.
+* For a relational database, to commit any pending transactions that were created by the sink data.
 
-A MESSAGE frame that contains a 2PC phase 1 request will specify the byte offset range(s) of the data sent to the sink for the logical time range that is managed by the Wallaroo global checkpoint protocol.
+If the Phase 2 decision is ABORT, then the connector sink must discard all application sink data in the transaction's byte range(s).
 
-The connector sink should be able to manage the durability of the data in this byte offset range(s).  If the connector sink cannot manage durability of this sink data, then the connector sink must always vote COMMIT during phase 1; such a sink then cannot provide the end-to-end message delivery guarantees that the connector sink protocol is intended to provide.  In such cases, we recommend that the Wallaroo TCPSink be used instead.
+#### 5.c. About aloc_sink's 2PC implementation
```

Comment:
Does `aloc_sink` mean "at least once sink"? If so, it should probably be named something like `eo_sink`, right?

Reply:
I'm not suggesting you make that change now, btw, but if this is so, we should open issues to rename.

Reply:
I think "aloc" stands for "at least once connector".  I don't remember its history. Perhaps Brian or Nissan coined it?